### PR TITLE
executor: Support NO_AUTO_CREATE_USER sql mode

### PR DESCRIPTION
**Pull request by morgo**
### What problem does this PR solve? 

Provides NO_AUTO_CREATE_USER sql mode.  Fixes  https://github.com/pingcap/tidb/issues/8128 

### What is changed and how it works?

It depends on parser changes:
https://github.com/pingcap/parser/pull/10

### Check List 

Tests 

 - Unit test
 - Manual test (add detailed scripts or steps below)

Code changes

 - Has exported function/method change
 - Has interface methods change
 - Has persistent data change

Side effects

 - Increased code complexity
 - Breaking backward compatibility (* when enabled by default: a followup patch)

Related changes

 - Need to be included in the release note

### CODE REVIEW

**Comment by morgo** [comment]
The build error is related to the parser changes not merged.  It should otherwise work.

**Comment by morgo** [comment]
PTAL @tiancaiamao

File: executor/grant.go [hunk 1]
```diff
@@ -38,6 +39,14 @@ var (
 	_ Executor = (*GrantExec)(nil)
 )
 
+const (
+	codePasswordNoMatch = 1133
```
Comment (XuHuaiyu) [inline executor/grant.go]: It better to put line42~line48 to executor/errors.go

File: executor/grant_test.go [hunk 2]
```diff
@@ -181,14 +181,25 @@ func (s *testSuite) TestIssue2456(c *C) {
 	tk.MustExec("GRANT ALL PRIVILEGES ON `dddb_%`.`te%` to 'dduser'@'%';")
 }
 
+func (s *testSuite) TestNoAutoCreateUser(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec(`DROP USER IF EXISTS 'test'@'%'`)
+	tk.MustExec(`SET sql_mode='NO_AUTO_CREATE_USER'`)
+	_, err := tk.Exec(`GRANT ALL PRIVILEGES ON *.* to 'test'@'%' IDENTIFIED BY 'xxx'`)
+	c.Check(err, NotNil)
```
Comment (XuHuaiyu) [inline executor/grant_test.go]: We'd better check the ErrorCode here.
https://github.com/pingcap/tidb/blob/master/ddl/db_test.go#L119-L126 can be referred.

**Comment by morgo** [comment]
Thx for the feedback - I've made the changes, but the CI will still show as 'X' because of the parser dependency.

**Comment by XuHuaiyu** [comment]
LGTM
PTAL @tiancaiamao

**Comment by tiancaiamao** [comment]
Try `GO111MODULE=on go get -u github.com/pingcap/parser` and `make dev` again
@morgo

**Comment by tiancaiamao** [comment]
Well, `GO111MODULE=on go get -u github.com/pingcap/parser` can't work temporarily.
Please wait for https://github.com/pingcap/parser/pull/23

**Pull request merged**